neuralpp/symbolic/normalizer.py

```diff
@@ -0,0 +1,67 @@
+from typing import Optional, Deque
+from collections import deque
+
+from .expression import Expression, Constant, FunctionApplication
+from .z3_expression import Z3SolverExpression
+from .constants import basic_true, basic_false, if_then_else
+from .context_simplifier import ContextSimplifier
+from .parameters import sympy_evaluate
+import neuralpp.symbolic.functions as functions
+
+
+def is_boolean_atom(expression: Expression) -> bool:
+    return expression.type == bool and not isinstance(expression, Constant)  # should ignore `True` and `False`
+
+
+def _bfs_first_atom_of_expression(expression: Expression, queue: Deque) -> Optional[Expression]:
+    queue.append(expression)
+    while queue:
+        current_expression = queue.popleft()
+        if is_boolean_atom(current_expression):
+            return current_expression
+        match current_expression:
+            case FunctionApplication(function=Constant(value=functions.conditional), arguments=arguments):
+                queue.extend(arguments[1:])  # do not include the if clause, it's meaningless
+            case FunctionApplication(arguments=arguments):
+                queue.extend(arguments)
+            case _:
+                pass  # atomic expression has no subexpressions
+    return None
+
+
+def first_atom_of_expression(expression: Expression) -> Optional[Expression]:
+    """ returns the first atom occurring in a given expression in a breadth-first search. """
```

Comment:
Probably worth mentioning that we're looking for the first atom *that's not a constant*

Reply:
I think the current definition of `atom` is not very clear as it can be "non-constant expression" (as used in `is_boolean_atom`), or "non-constant boolean expression" (as used in `first_atom_of_expression`), or "boolean expression" (as proposed by you here, if we change the function name to `first_non_constant_atom_of_expression`). 

Also, since "atom" is not a very straight-forward name, we've switched to "formula" in the project plan. I think we should do the same here. "Formula" is a better understood name, which should mean "boolean expression".